src/coap_gnutls.c

```diff
@@ -770,6 +774,98 @@ setup_pki_credentials(gnutls_certificate_credentials_t *pki_credentials,
     }
     break;
 
+  case COAP_PKI_KEY_PEM_BUF:
+    if (setup_data->pki_key.key.pem_buf.public_cert &&
+        setup_data->pki_key.key.pem_buf.public_cert_len &&
+        setup_data->pki_key.key.pem_buf.private_key &&
+        setup_data->pki_key.key.pem_buf.private_key_len) {
+      gnutls_datum_t cert;
+      gnutls_datum_t key;
+      int alloced_cert_memory = 0;
+      int alloced_key_memory = 0;
+
+      cert.size = setup_data->pki_key.key.pem_buf.public_cert_len;
+      if (setup_data->pki_key.key.pem_buf.public_cert[cert.size-1] != '\000') {
+        /* Need to allocate memory, rather than just copying pointers across */
+        alloced_cert_memory = 1;
+        cert.data = gnutls_malloc(cert.size + 1);
+        if (!cert.data) {
+          coap_log(LOG_ERR, "gnutls_malloc failure\n");
+          return GNUTLS_E_MEMORY_ERROR;
+        }
+        memcpy(cert.data, setup_data->pki_key.key.pem_buf.public_cert,
+               cert.size);
+        cert.data[cert.size] = '\000';
+        cert.size++;
+      }
+      else {
+        /* To get around const issue */
+        memcpy(&cert.data,
+             &setup_data->pki_key.key.pem_buf.public_cert, sizeof(cert.data));
+      }
+      key.size = setup_data->pki_key.key.pem_buf.private_key_len;
+      if (setup_data->pki_key.key.pem_buf.private_key[key.size-1] != '\000') {
+        /* Need to allocate memory, rather than just copying pointers across */
+        alloced_key_memory = 1;
+        key.data = gnutls_malloc(key.size + 1);
+        if (!key.data) {
+          coap_log(LOG_ERR, "gnutls_malloc failure\n");
+          return GNUTLS_E_MEMORY_ERROR;
+        }
+        memcpy(key.data, setup_data->pki_key.key.pem_buf.private_key, key.size);
+        key.data[key.size] = '\000';
+        key.size++;
+      }
+      else {
+        /* To get around const issue */
+        memcpy(&key.data,
+               &setup_data->pki_key.key.pem_buf.private_key, sizeof(key.data));
+      }
+      G_CHECK(gnutls_certificate_set_x509_key_mem(*pki_credentials,
+                                   &cert,
+                                   &key,
+                                   GNUTLS_X509_FMT_PEM),
+                 "gnutls_certificate_set_x509_key_mem");
+      if (alloced_cert_memory) gnutls_free(cert.data);
+      if (alloced_key_memory) gnutls_free(key.data);
+    }
+    else if (role == COAP_DTLS_ROLE_SERVER) {
+      coap_log(LOG_ERR,
+               "***setup_pki: (D)TLS: No Server Certificate + Private "
+               "Key defined\n");
+      return GNUTLS_E_INSUFFICIENT_CREDENTIALS;
+    }
+    if (setup_data->pki_key.key.pem_buf.ca_cert &&
+        setup_data->pki_key.key.pem_buf.ca_cert_len) {
+      gnutls_datum_t ca;
+      int alloced_ca_memory = 0;
+
+      ca.size = setup_data->pki_key.key.pem_buf.ca_cert_len;
+      if (setup_data->pki_key.key.pem_buf.ca_cert[ca.size-1] != '\000') {
```

Comment:
Here also: `ca.size` may be 0.

Reply:
Again caught by the code above testing setup_data->pki_key.key.pem_buf.ca_cert_len
````
    if (setup_data->pki_key.key.pem_buf.ca_cert &&
        setup_data->pki_key.key.pem_buf.ca_cert_len) {
      gnutls_datum_t ca;
      int alloced_ca_memory = 0;
````
so never can be 0.